cohere/__init__.py

```diff
@@ -1,8 +1,12 @@
-from .client import Client
-from .error import CohereError
+from cohere.client import Client
+from cohere.client_async import AsyncClient
+from cohere.error import CohereError, CohereAPIError,CohereConnectionError
 
 COHERE_API_URL = 'https://api.cohere.ai'
-COHERE_VERSION = '2022-12-06'
+RETRY_STATUS_CODES = [429, 500, 502, 503, 504]
```

Comment:
does the retry policy handle 429 correctly? You need to wait for one minute before retrying i believe

Reply:
It does the same thing on all those error codes. Where is the 429 behaviour documented (preferably publically)

Reply:
I think we need to handle it separately:

```
if response.status == 429:
  sleep(60)
  make_request()
```

Unfortunately we don't have public documentation for the rate limiting. I just requested it again: https://cohereai.slack.com/archives/C02L95WS1B2/p1678767941930239

The way it works is that depending on what type of API key you are using and what endpoint you are hitting you are allowed X number of requests per minute, if you exceed that amount, you start receiving 429 status code and need to wait for 1 minute before you can make more requests.

Here are the limits per endpoint / key types: https://github.com/cohere-ai/blobheart/blob/main/config/limits/limits.go

Reply:
> and need to wait for 1 minute before you can make more requests.

This is what I'm not sure about. I thought if you have a limit of 100/min, and you hit the limit at 50s by making 120 calls/min, you can continue making calls by whenever your rolling window of the last 60s doesn't have 100 calls -- which is in like 1-2s, not 60s.
I know there are headers which suggest when to make the next call, but these assume no concurrency I believe.

In any case I think the asyncio client here is at feature parity with the sync one, that is, they both don't handle this perfectly.

Reply:
added sleep 5